pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -376,6 +464,53 @@ func (b *Base) Objects(rsrc interface{}, rsrclabels map[string]string, observed,
 		}
 	}
 
+	item := k8s.GetItem(observed, &batchv1.Job{}, getUpgradeJobName(master), "default")
+	if item == nil {
+	}
+	var upgradeJob *batchv1.Job
+	if item != nil {
+		upgradeJob = item.(*batchv1.Job)
+	}
+	var upgradeResources []reconciler.Object
+	resetImage := true
+	if upgradeJob == nil && !master.Status.Meta.IsConditionTrue(upgradeFailed) &&
+		master.Status.ImageToUse != "" && master.Status.ImageToUse != master.Spec.Image {
+		var err error
+		upgradeResources, err = launchJob(master, rsrclabels)
+		if err != nil {
+			return nil, err
+		}
+		resetImage = false
+	} else if upgradeJob != nil && upgradeJob.Status.Succeeded == 0 && upgradeJob.Status.Failed < upgradeFailureLimit {
+		var err error
+		upgradeResources, err = getPreUpgradeJobResources(master, rsrclabels)
+		if err != nil {
+			return nil, err
+		}
+		resetImage = false
+	} else if upgradeJob != nil && upgradeJob.Status.Succeeded > 0 {
+		master.Status.ImageToUse = master.Spec.Image
+		master.Status.UserInterfaceImageToUse = master.Spec.UserInterfaceImage
+		master.Status.UpgradeJobVersion = uuid.New().String()
```

Comment:
Why generate a new uuid?

Reply:
Prevent the scenario where we finish the whole process and start a new upgrade before the job is removed by ttlSecondsAfterFinished. Then, the controller will see that upgradeJob != nil and not start a new upgrade.